std/socket.zig

```diff
@@ -0,0 +1,678 @@
+                system.EINVAL => unreachable,
+                system.ENOPROTOOPT => unreachable, // The option is unknown at the level indicated.
+                system.ENOTSOCK => unreachable, // The file descriptor socket does not refer to a socket.
+            }
+        } else {
+            const err = system.WSAGetLastError();
+            switch (err) {
+                0 => switch (err_code) {
+                    0 => return,
+                    system.ERROR.WSAEACCES => return ConnectError.PermissionDenied,
+                    system.ERROR.WSAEADDRINUSE => return ConnectError.AddressInUse,
+                    system.ERROR.WSAEADDRNOTAVAIL => return ConnectError.AddressNotAvailable,
+                    system.ERROR.WSAEAFNOSUPPORT => return ConnectError.AddressFamilyNotSupported,
+                    system.ERROR.WSAEWOULDBLOCK => return ConnectError.system.emResources,
+                    system.ERROR.WSAEALREADY => unreachable, // The socket is nonblocking and a previous connection attempt has not yet been completed.
+                    system.ERROR.WSAEBADF => unreachable, // socket is not a valid open file descriptor.
+                    system.ERROR.WSAECONNREFUSED => return ConnectError.ConnectionRefused,
+                    system.ERROR.WSAEFAULT => unreachable, // The socket structure address is outside the user's address space.
+                    system.ERROR.WSAEISCONN => unreachable, // The socket is already connected.
+                    system.ERROR.WSAENETUNREACH => return ConnectError.NetworkUnreachable,
+                    system.ERROR.WSAENOTSOCK => unreachable, // The file descriptor socket does not refer to a socket.
+                    system.ERROR.WSAEPROTOTYPE => unreachable, // The socket type does not support the requested communications protocol.
+                    system.ERROR.WSAETIMEDOUT => return ConnectError.ConnectionTimedOut,
+                    else => return unexpectedError(@intCast(u32, err)),
+                },
+                else => return unexpectedError(@intCast(u32, err)),
+                system.ERROR.WSAEBADF => unreachable, // The argument socket is not a valid file descriptor.
+                system.ERROR.WSAEFAULT => unreachable, // The address pointed to by optval or optlen is not in a valid part of the process address space.
+                system.ERROR.WSAEINVAL => unreachable,
+                system.ERROR.WSAENOPROTOOPT => unreachable, // The option is unknown at the level indicated.
+                system.ERROR.WSAENOTSOCK => unreachable, // The file descriptor socket does not refer to a socket.
+            }
+        }
+    }
+
+    pub fn getSockName(self: Socket) GetSockNameError!net.Address {
+        var addr: net.OsAddress = undefined;
+        var addrlen: system.socklen_t = @sizeOf(net.OsAddress);
+        const rc = system.getsockname(self.fd, &addr, &addrlen);
+
+        if (is_posix) {
+            const err = system.getErrno(rc);
```

Comment:
In general you shouldn't be checking the errno unless the return value of the function indicates an error occured.

The errno can be set in non-error cases and can also be set to a non-zero value by other functions before this function is called.

Reply:
Here's a link with more information on it: https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152351

Reply:
You're thinking in C :)

Zig follows linux in that errnos are returned from the system library as negative values.
To check for errors you have to extract the errno and ensure it's 0.

Reply:
Oh I see that the posix version is using the return value which is fine.  It's the Windows version that is a problem here.  WSAGetLastError should ONLY ever be called if the return value of the function you called indicated there was an error.  Otherwise, the errno could already have been set to a non-zero value before this function was called.  For example, if someone called this function, it failed, then it was called again later, the errno would be non-zero at that point even if it was succesful the second time.  The errno is not normally initialized back to zero every time it is checked or inside every function, this is because developer's don't need to  initialize it because it should only be checked if the function indicated there was an error.